geth/node/syncpoll.go

```diff
@@ -0,0 +1,95 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/les"
+)
+
+// errors
+var (
+	ErrNodeSyncFailedToStart = errors.New("node synchronization failed to start")
+	ErrNodeSyncTakesTooLong  = errors.New("node synchronization is taking too long")
+)
+
+// delays
+var (
+	// delayCycleForSyncStart sets the timeout for checking state of synchronization starting.
+	delayCycleForSyncStart = 100 * time.Millisecond
+)
+
+// SyncPoll provides a structure that allows us to check the status of
+// ethereum node synchronization.
+type SyncPoll struct {
+	eth *les.LightEthereum
+}
+
+// NewSyncPoll returns a new instance of SyncPoll.
+func NewSyncPoll(leth *les.LightEthereum) *SyncPoll {
+	return &SyncPoll{
+		eth: leth,
+	}
+}
+
+// Poll returns a channel which allows the user to listen for a done signal
+// as to when the node has finished syncing or stop due to an error.
+func (n *SyncPoll) Poll(ctx context.Context) error {
+	errChan := make(chan error)
+	downloader := n.eth.Downloader()
+
+	syncStart := make(chan struct{})
+	go n.pollSyncStart(ctx, syncStart, errChan, downloader)
```

Comment:
Running a goroutine with immediate `select` statement can be written without both of them:

```go
if err := n.pollSyncStart(ctx); err != nil {
  return err
}

if err := n.pollSyncCompleted(ctx); err != nil {
  return err
}

return nil
```

Also, we could have simpler methods:
```go
func (n *SyncPoll) pollSyncStart(ctx context.Context) error {
  for {
    select {
    case <-ctx.Done():
      return ErrNodeSyncFailedToStart
    case <-time.After(delayCycleForSyncStart):
        if n.downloader.Synchronising() {
          return nil
        }
  }
}
```

Reply:
Indeed, I think my brain shortcircuit, channels must have caused this. Thanks for the tip.